test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Audience.cs

```diff
@@ -0,0 +1,215 @@
+                    tokenValidationParametersResult.Claims,
+                    validationParametersResult.UnwrapResult().Claims,
+                    context);
+            }
+            else
+            {
+                theoryData.ExpectedException.ProcessException(tokenValidationParametersResult.Exception, context);
+
+                if (!validationParametersResult.IsSuccess)
+                {
+                    // If there is a special case for the ValidationParameters path, use that.
+                    if (theoryData.ExpectedExceptionValidationParameters != null)
+                        theoryData.ExpectedExceptionValidationParameters
+                            .ProcessException(validationParametersResult.UnwrapError().GetException(), context);
+                    else
+                        theoryData.ExpectedException
+                            .ProcessException(validationParametersResult.UnwrapError().GetException(), context);
+                }
+            }
+
+            TestUtilities.AssertFailIfErrors(context);
+        }
+
+        public static TheoryData<ValidateTokenAsyncAudienceTheoryData> ValidateTokenAsync_AudienceTestCases
+        {
+            get
+            {
+                return new TheoryData<ValidateTokenAsyncAudienceTheoryData>
+                {
+                    new ValidateTokenAsyncAudienceTheoryData("Valid_AudiencesMatch")
+                    {
+                        Audience = Default.Audience,
+                        TokenValidationParameters = CreateTokenValidationParameters([Default.Audience]),
+                        ValidationParameters = CreateValidationParameters([Default.Audience]),
+                    },
+                    new ValidateTokenAsyncAudienceTheoryData("Invalid_AudiencesDontMatch")
+                    {
+                        // This scenario is the same if the token audience is an empty string or whitespace.
+                        // As long as the token audience and the valid audience are not equal, the validation fails.
+                        TokenValidationParameters = CreateTokenValidationParameters([Default.Audience]),
+                        ValidationParameters = CreateValidationParameters([Default.Audience]),
+                        Audience = "InvalidAudience",
+                        ExpectedIsValid = false,
+                        ExpectedException = ExpectedException.SecurityTokenInvalidAudienceException("IDX10214:"),
+                        // ValidateTokenAsync with ValidationParameters returns a different error message to account for the
+                        // removal of the ValidAudience property from the ValidationParameters class.
+                        ExpectedExceptionValidationParameters = ExpectedException.SecurityTokenInvalidAudienceException("IDX10215:"),
+                    },
+                    new ValidateTokenAsyncAudienceTheoryData("Valid_AudienceWithinValidAudiences")
+                    {
```

Comment:
there is a lot of duplicate code, which makes it harder to review and understand, imho. Not sure of a better solution though.

Reply:
Originally we were using a single test class and have all the theory data alongside, but it was requested by Brent that we decouple the different aspects to reduce the size of the files and make it easier to follow what aspects of the validation are being tested.

In order to separate this, we have to duplicate some of the test scenario code. At the end of the day, the tests are doing the same steps:
- Create the token based on some configuration
- Validate the created token using the legacy validation path
- Validate the created token using the new path
- Compare the result from one against the other
- Compare the claims and claims identify from both so that they are identical.
- Ensure the expected validity and exception match

Reply:
Extracted the duplicated code to a common file.